src/main/scala/SchedulerImp.scala

```diff
@@ -0,0 +1,38 @@
+import java.util.concurrent.PriorityBlockingQueue
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SchedulerImp(
+    val minHeap: PriorityBlockingQueue[ScheduledTask]
+) extends Scheduler {
+
+  @volatile var running = false
+
+  override def schedule(task: () => Unit, timestamp: Long): Unit = {
+    minHeap.offer(ScheduledTask(task, timestamp))
+  }
+
+  override def start()(implicit ec: ExecutionContext): Unit = {
+    if (!running) {
+      running = true
+      Future {
+        while (running) {
```

Comment:
because only the main thread writes the `running` variable and the worker threads only read it, we don't need intrinsic lock for the `running`. but, use `volatile` to ensure it read/write to main memory instead of CPU cache.